sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridConsumer.cs

```diff
@@ -0,0 +1,301 @@
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventGridConsumer"/> class.
+        /// </summary>
+        /// <param name="options"> Options for configuring deserialization. </param>
+        public EventGridConsumer(EventGridConsumerOptions options)
+        {
+            Argument.AssertNotNull(options, nameof(options));
+            _objectSerializer = options.DataSerializer;
+            _customEventTypeMappings = options.CustomEventTypeMappings;
+        }
+
+        /// <summary>
+        /// Deserializes JSON encoded events and returns an array of events encoded in the EventGrid event schema.
+        /// </summary>
+        /// <param name="requestContent">
+        /// The JSON encoded representation of either a single event or an array or events, encoded in the EventGrid event schema.
+        /// </param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns>A list of EventGrid Events.</returns>
+        public virtual EventGridEvent[] DeserializeEventGridEvents(string requestContent, CancellationToken cancellationToken = default)
+            => DeserializeEventGridEventsInternal(requestContent, false /*async*/, cancellationToken).EnsureCompleted();
+
+        /// <summary>
+        /// Deserializes JSON encoded events and returns an array of events encoded in the EventGrid event schema.
+        /// </summary>
+        /// <param name="requestContent">
+        /// The JSON encoded representation of either a single event or an array or events, encoded in the EventGrid event schema.
+        /// </param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <returns>A list of EventGrid Events.</returns>
+        public virtual async Task<EventGridEvent[]> DeserializeEventGridEventsAsync(string requestContent, CancellationToken cancellationToken = default)
+            => await DeserializeEventGridEventsInternal(requestContent, true /*async*/, cancellationToken).ConfigureAwait(false);
+
+        private async Task<EventGridEvent[]> DeserializeEventGridEventsInternal(string requestContent, bool async, CancellationToken cancellationToken = default)
+        {
+            List<EventGridEventInternal> egInternalEvents = new List<EventGridEventInternal>();
+            List<EventGridEvent> egEvents = new List<EventGridEvent>();
+
+            // Deserialize raw JSON string into separate events, deserialize event envelope properties
+            MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes(requestContent));
+            JsonDocument requestDocument;
+            if (async)
+            {
+                requestDocument = await JsonDocument.ParseAsync(stream, default, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                requestDocument = JsonDocument.Parse(stream, default);
+            }
```

Comment:
nit: factor this out into a method since we also use it in the other Deserialize method.